### PR TITLE
Corrected BoundingBox.relate Within case

### PR DESCRIPTION
**Pull request by zebehringer**
use [new] Polygon:line contains test instead of intersects
optimized the code a little to avoid extra intersection test

### CODE REVIEW

File: src/main/scala/magellan/BoundingBox.scala [hunk 1]
```diff
@@ -112,14 +112,18 @@ case class BoundingBox(xmin: Double, ymin: Double, xmax: Double, ymax: Double) {
...
     )
 
-    val lineIntersections = (lines filter (shape intersects _)).size
+    if (contains(shape.boundingBox)) {
+      return Contains
+    }
+
+    val linesContained = (lines filter (shape contains _)).size
     val vertexContained = (vertices filter (shape contains _)).size
+    if (linesContained == lines.length && vertexContained == 4) {
```
Comment (zebehringer) [inline src/main/scala/magellan/BoundingBox.scala]: actually, ```vertexContained == 4``` is probably not needed here

File: src/main/scala/magellan/Polygon.scala [hunk 2]
```diff
@@ -169,6 +169,20 @@ class Polygon extends Shape {
...
+  /**
+    * A polygon wholly contains a line if none of its edges intersect the line
+     AND contains one of the line endpoints.
+    *
+    * @param line
+    * @return
+    */
+  private [magellan] def contains(line: Line): Boolean = {
+    if (loops.exists(_.intersects(line))) {
+      false
```
Comment (zebehringer) [inline src/main/scala/magellan/Polygon.scala]: I think I need a return here

**Comment by harsha2010** [comment]
@zebehringer as discussed offline, thanks a lot for catching this performance regression.
I will take your PR and add some tests to it to capture the regression and then merge this PR in today

File: src/main/scala/magellan/Polygon.scala [hunk 3]
```diff
@@ -169,6 +169,20 @@ class Polygon extends Shape {
...
   }
 
+  /**
+    * A polygon wholly contains a line if none of its edges intersect the line
+     AND contains one of the line endpoints.
+    *
+    * @param line
+    * @return
+    */
+  private [magellan] def contains(line: Line): Boolean = {
```
Comment (harsha2010) [inline src/main/scala/magellan/Polygon.scala]: contains does not mean wholly contains... a polygon contains a line even if the edges are touching as long as the interior is fully contained. you probably want to introduce the concept of strict containment here

Comment (zebehringer) [reply src/main/scala/magellan/Polygon.scala]: I see. It is getting complicated

File: src/main/scala/magellan/BoundingBox.scala [hunk 4]
```diff
@@ -112,14 +112,18 @@ case class BoundingBox(xmin: Double, ymin: Double, xmax: Double, ymax: Double) {
       Line(Point(xmin, ymax), Point(xmax, ymin))
     )
 
+    if (contains(shape.boundingBox)) {
+      return Contains
```
Comment (zebehringer) [inline src/main/scala/magellan/BoundingBox.scala]: somewhat related, but am I correct that this Relation is ignored by the shortcut?

File: src/main/scala/magellan/Polygon.scala (hunk 3 above)
Comment (zebehringer) [reply src/main/scala/magellan/Polygon.scala]: would the equivalent PostGIS concept be ST_Covers?

**Comment by harsha2010** [comment]
@zebehringer I am closing this, I have created a PR #207  based off this which adds some tests and missing functionality as well. Please review it when you get a chance and if possible test it out
Thanks for noticing this issue!

**Pull request closed without merging**